tlatools/org.lamport.tlatools/test/pcal/parser/PcalParserTranslator.java

```diff
@@ -0,0 +1,184 @@
+	 * extents then providing a list of lines of tokens. So we keep TLA+
+	 * expressions quite simple here. It could be interesting in the future
+	 * to call into SANY to parse these, so we can analyze the logic that
+	 * determines expression extent.
+	 * 
+	 * @param expr The TLA+ expression, as a sequence of lines of tokens.
+	 * @return The corresponding node in normalized form.
+	 */
+	public static AstNode translate(TLAExpr expr) {
+		Assert.assertEquals(1, expr.tokens.size()); // Lines of tokens
+		Assert.assertEquals(1, expr.tokens.get(0).size()); // Tokens on a given line
+		return translate(expr.tokens.get(0).get(0));
+	}
+	
+	/**
+	 * Translates the PlusCal AST generated by the parser into a normalized
+	 * form for comparison to the unified parser test corpus.
+	 * 
+	 * TODO: This translation is largely incomplete in its current form. The
+	 * bones are now in place for a comprehensive PlusCal syntax corpus test
+	 * similar to what exists for SANY, but now somebody needs to do the real
+	 * work of translating every syntax node to its universal corpus form.
+	 * 
+	 * @param astNode A node in the PlusCal AST.
+	 * @return The corresponding node in normalized universal corpus form.
+	 * @throws ParseAlgorithmException If a construct could not be translated.
+	 */
+	public static AstNode translate(AST astNode) throws ParseAlgorithmException {
+		if (astNode.getClass().equals(AST.UniprocessObj.getClass())) {
+			AST.Uniprocess node = (AST.Uniprocess)astNode;
+			AstNode sourceFile = Kind.SOURCE_FILE.asNode();
+			AstNode pcalAlgorithm = Kind.PCAL_ALGORITHM.asNode();
+			AstNode algorithmStart = Kind.PCAL_ALGORITHM_START.asNode();
+			if (PcalParams.FairAlgorithm) {
+				algorithmStart.addChild(Kind.FAIR.asNode());
+			}
+			pcalAlgorithm.addChild(algorithmStart);
+			Assert.assertNotNull(node.name);
+			Assert.assertFalse(node.name.trim().isEmpty());
+			pcalAlgorithm.addField("name", Kind.IDENTIFIER.asNode());
+			Assert.assertNotEquals(0, node.body.size());
+			AstNode algorithmBody = Kind.PCAL_ALGORITHM_BODY.asNode();
+			for (Object child : node.body) {
+				algorithmBody.addChild(translate((AST)child));
+			}
+			pcalAlgorithm.addChild(algorithmBody);
+			sourceFile.addChild(pcalAlgorithm);
+			return sourceFile;
+		} else if (astNode.getClass().equals(AST.LabeledStmtObj.getClass())) {
+			AST.LabeledStmt node = (AST.LabeledStmt)astNode;
+			Assert.assertEquals(1, node.stmts.size());
+			return translate((AST)node.stmts.get(0));
+		} else if (astNode.getClass().equals(AST.SkipObj.getClass())) {
+			return Kind.PCAL_SKIP.asNode();
+		} else if (astNode.getClass().equals(AST.AssertObj.getClass())) {
+			AST.Assert node = (AST.Assert)astNode;
+			AstNode assertNode = Kind.PCAL_ASSERT.asNode();
+			assertNode.addChild(translate(node.exp));
+			return assertNode;
+		} else {
+			throw new ParseAlgorithmException("Unhandled AST class " + astNode.getClass().toString());
```

Comment:
Nit: `Class.toString()` already includes the word "class":
```
jshell> String.class.toString()
$1 ==> "class java.lang.String"
```
I would replace `toString` with `getCanonicalName` here.